bin/check-package-metadata.js

```diff
@@ -0,0 +1,192 @@
+// License text available at https://opensource.org/licenses/MIT
+
+/**
+ * This is an internal script that will verify that all packages in our monorepo
+ * are registered in all required places and have the expected metadata in their
+ * package.json.
+ */
+'use strict';
+
+const path = require('path');
+const fs = require('fs-extra');
+
+const Project = require('@lerna/project');
+
+/**
+ * Check existence of LICENSE file in the monorepo packages
+ * @param {Package[]} packages A list of @lerna/project packages
+ */
+async function checkLicenseFiles(packages) {
+  const errors = [];
+
+  for await (const p of packages) {
+    const exists = await fs.pathExists(path.join(p.location, 'LICENSE'));
+
+    if (!exists) {
+      errors.push(`${p.name} directory doesn't contain LICENSE file`);
+    }
+  }
+
+  return errors;
+}
+
+/**
+ * Check mention of packages in MONOREPO.md file
+ * @param {Package[]} packages A list of @lerna/project packages
+ * @param {string} rootPath A project.rootPath
+ */
+async function checkMonorepoFile(packages, rootPath) {
+  const errors = [];
+  const monorepoFile = await fs.readFile('docs/site/MONOREPO.md', 'utf8');
+
+  for (const p of packages) {
+    const packagePath = path.relative(rootPath, p.location);
+
+    const packageExists =
+      monorepoFile.includes(p.name) || monorepoFile.includes(packagePath);
+
+    if (!packageExists) {
+      errors.push(`${p.name} is not added in the MONOREPO.md file`);
+    }
+  }
+
+  return errors;
+}
+
+/**
+ * Check mention of packages in CODEOWNERS.md file
+ * @param {Package[]} packages A list of @lerna/project packages
+ * @param {string} rootPath A project.rootPath
+ */
+async function checkCodeOwnersFile(packages, rootPath) {
+  const errors = [];
+  const excludes = ['@loopback/sandbox-example'];
+
+  const codeOwnersFile = await fs.readFile('CODEOWNERS', 'utf8');
+
+  for (const p of packages) {
+    if (excludes.includes(p.name)) {
+      continue;
+    }
+
+    const packagePath = path.relative(rootPath, p.location);
+
+    if (!codeOwnersFile.includes(packagePath)) {
+      errors.push(`${p.name} is not added in the CODEOWNERS.md file`);
+    }
+  }
+
+  return errors;
+}
+
+/**
+ * Check all required fields of package.json for each package on the matching with root package.json
+ * @param {Package[]} packages A list of @lerna/project packages
+ * @param {Object} rootPkg A root package.json
+ */
+async function checkPkgsPackageJson(packages, rootPkg) {
+  const errors = [];
+  const pkgs = packages.filter(p => !p.private);
+
+  function getCopyrightOwner(pkg) {
+    return pkg['copyright.owner'] || (pkg.copyright && pkg.copyright.owner);
+  }
+
+  function getErrorText(packageName, field) {
+    return `${packageName} package.json doesn't contain ${field} field or has incorrect value`;
+  }
+
+  for (const p of pkgs) {
+    const pkg = fs.readJsonSync(path.join(p.location, 'package.json'));
+
+    const isPublicAccess =
```

Comment:
Let's also check the following properties:

```json
{
  "main": "dist/index.js", 
  "types": "dist/index.d.ts"
}
```

Reply:
@raymondfeng , I'd checked the field `main` and `types` in the packages, and noticed the packages: `@loopback/docs`, `@loopback/build`, `@loopback/cli`, `@loopback/eslint-config` should be excluded from `checkPkgsPackageJson` method. I found it's ugly to add this kind of checks in the `checkPkgsPackageJson` method. I offer to standardize validation process:

```js
function defaultErrorText(packet, field) {
  return `${packet.name} package.json doesn't contain ${field} field or has incorrect value`;
}

function getPkgField(pkg, field) {
  return (
    pkg[field] ||
    field.split('.').reduce((accumulator, part) => {
      return accumulator && accumulator[part];
    }, pkg)
  );
}

function applyRule(packet, pkg, field, rule) {
  if (typeof rule === 'function') {
    return rule(packet, pkg);
  }

  return rule === getPkgField(pkg, field);
}

function applyRules(packet, pkg, field, rules) {
  if (!Array.isArray(rules)) {
    rules = [rules];
  }

  return rules.reduce(
    (accumulator, {rule, exclude = [], getErrorText = defaultErrorText}) => {
      if (exclude.includes(packet.name)) {
        return accumulator;
      }

      if (!applyRule(packet, pkg, field, rule)) {
        accumulator.push(getErrorText(packet, field));
      }

      return accumulator;
    },
    [],
  );
}

function applySchema(packet, schema) {
  const errors = [];
  const pkg = fs.readJsonSync(path.join(packet.location, 'package.json'));

  for (const field in schema) {
    errors.push(...applyRules(packet, pkg, field, schema[field]));
  }

  return errors;
}

async function checkPkgsPackageJson(packages, rootPkg) {
  const SCHEMA = {
    author: {
      rule: rootPkg.author,
    },
    'copyright.owner': {
      rule: getCopyrightOwner(rootPkg),
    },
    license: {
      rule: rootPkg.license,
    },
    'publishConfig.access': {
      rule: 'public',
    },
    main: {
      rule: 'dist/index.js',
      exclude: [
        '@loopback/docs',
        '@loopback/build',
        '@loopback/cli',
        '@loopback/eslint-config',
      ],
    },
    types: {
      rule: 'dist/index.d.ts',
      exclude: [
        '@loopback/docs',
        '@loopback/build',
        '@loopback/cli',
        '@loopback/eslint-config',
      ],
    },
    'repository.url': {
      rule: 'https://github.com/strongloop/loopback-next.git',
    },
    'repository.directory': [
      {
        rule: (_, pkg) => pkg.repository.directory,
      },
      {
        rule: (_, pkg) => fs.existsSync(pkg.repository.directory),
        getErrorText: p => `${p.name} directory doesn't exist in the monorepo`,
      },
    ],
  };

  const errors = [];
  const pkgs = packages.filter(p => !p.private);

  function getCopyrightOwner(pkg) {
    return pkg['copyright.owner'] || (pkg.copyright && pkg.copyright.owner);
  }

  for (const p of pkgs) {
    errors.push(...applySchema(p, SCHEMA));
  }

  return errors;
}
```

I guess it's easy to support and easy to add new rules. But maybe it's overengineering and doesn't make sense. Anyway, please give your feedback about my suggestion.

Reply:
Haha, it is actually simpler. Those packages are not TypeScript modules. We can check the presence of tsconfig.json.

Reply:
OMG) Ok, thank you)